doc/nvim-surround.txt

```diff
@@ -1,357 +1,529 @@
-            value of each mapping determines whether the character is used to
-            change just the type of the tag, or the whole tag contents.
-
-            Default: {
-                ["t"] = "type",
-                ["T"] = "whole",
-            }
-
-                                            *nvim-surround.delimiters.aliases*
-        aliases:~
-            A table that aliases some characters to other values:
-            * If the value is a single character, then the characters are
-              interchangeable.
-            * If the value is a table, then the alias may stand in for any of
-              the characters in the table when changing/deleting.
-
-            For example, `dsq` deletes the "closest" pair between `ds'`,
-            `ds"`, and `ds``. When typed at position 1, the double quotes `"`
-            get deleted, and when typed at position 2, the single quotes `'`
-            get deleted.
+        The user is queried for the function name, and that is used to create
+        the surrounding pair at runtime. For "static" surrounds that are not
+        evaluated at runtime, the function can be omitted and the `add` key
+        can directly take on the value of the delimiter pair. Furthermore,
+        tables with just one string can be replaced by the string itself. For
+        example, the following configuration also adds parentheses:
 >
-            Cursor:                      2             1
-            Buffer: vim.fn.execute("echo 'Hello world!'")
+        [")"] = {
+            add = { "(", ")" },
+        }
+<
+    find: ~
+        A function that returns the "parent selection" for a surrounding pair.
+        The selection returned should contain the surrounding pair and little
+        else, as it is used by the `delete` and `change` keys to determine
+        what to modify.
+
+    delete: ~
+        A function that returns the pair of selections to remove from the
+        buffer when deleting the surrounding pair.
+
+    change: ~
+        A table of two keys: `target` and `replacement`.
+
+        target: ~
+            A function that returns the pair of selections to be replaced in
+            the buffer when changing the surrounding pair. Similar to the
+            `delete` key.
+        replacement: ~
+            A function that returns the surrounding pair to replace the target
+            selections. Similar to the `add` key.
+
+    Note: When the `change.replacement` key is provided, the user is not
+    queried for a replacement character; the results of `change.replacement`
+    are directly used as the replacement pair. For example, when changing HTML
+    tag types, only `cst` is needed, instead of `cstt`.
+
+There is a special key in the `delimiters` table, `invalid_key_behavior`, that
+defines how |nvim-surround| behaves when a given key in the delimiter table is
+omitted. It takes the same form as all other delimiter tables. The default
```

Comment:
Personally, I think a better default is to simply delete / change the character specified. At least that's what I would expect as a user coming from other surround plugins. Curious what other users think here (cc @andrewferrier, @jonatan-branting, @akinsho).

Reply:
For some reason I was under the impression that `vim-surround` NOOP'd for everything; I've personally been using a setup that does just that. I'll probably change the default behavior for this then.

Reply:
Changed by 502a430.

Reply:
I just tested my claim with `vim-sandwich` and `vim-surround` and I must say I wasn't completely correct, sorry! `vim-sandwich` works as I have described but `vim-surround` doesn't: e.g. for `dsx` nothing happens but for non-letter characters such as `*` the characters are deleted. There seem to be special cases though such as:

- `*word*`, action: `ds*` => result: `word`
```
*
word
*
```
action `ds*` => result: nothing changed

Reply:
I think the current setup is easy enough to understand; I don't really want to bother with adding more special cases to handle everything. I don't think it's *that* hard to disable the built-in behavior and add your own keys into the delimiter table for `*`, `$`, `-`, etc.

Reply:
Yes of course, I always appreciate extra feedback. Also thank you so much for taking the time to review the code! I *still* haven't gotten around to writing a short blurb that introduces the code structure, so I apologize for that :smile: